cpp/src/arrow/c/abi.h

```diff
@@ -106,6 +169,77 @@ struct ArrowArrayStream {
 
 #endif  // ARROW_C_STREAM_INTERFACE
 
+#ifndef ARROW_C_DEVICE_STREAM_INTERFACE
+#define ARROW_C_DEVICE_STREAM_INTERFACE
+
+struct ArrowDeviceArrayStream {
+  // The device that this stream produces data on.
+  // All ArrowDeviceArrays that are produced by this
+  // stream should have the same device_type as set
+  // here. The device_type needs to be provided here
+  // so that consumers can provide the correct type
+  // of stream_ptr when calling get_next.
+  ArrowDeviceType device_type;
+
+  // Callback to get the stream schema
+  // (will be the same for all arrays in the stream).
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // If successful, the ArrowSchema must be released independently from the stream.
+  int (*get_schema)(struct ArrowDeviceArrayStream*, struct ArrowSchema* out);
+
+  // Callback to get the device id for the next array.
+  // This is necessary so that the proper/correct stream pointer can be provided
+  // to get_next. The parameter provided must not be null.
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // The next call to `get_next` should provide an ArrowDeviceArray whose
+  // device_id matches what is provided here, and whose device_type is the
+  // same as the device_type member of this stream.
+  int (*get_next_device_id)(struct ArrowDeviceArrayStream*, int* out_device_id);
+
+  // Callback to get the next array
+  // (if no error and the array is released, the stream has ended)
+  //
+  // the provided stream_ptr should be the appropriate stream, or
```

Comment:
to be clear, this is "CUDA stream"?

while this is a draft, the language here is a little imprecise/feels like it assumes too much knowledge of a particular programming model. It would be great to define any terminology needed so that things can reasonably stand alone.

Reply:
This is referring to whichever type is used by the appropriate backend to represent a stream (or should be null if the backend doesn't utilize this concept). 

I'll update the comments to be more explicit about this.

Reply:
I agree with Antoine's comment above that we need a different term for this in this context. But my general point is that it would help to have a set of definitions up front. As it stands this seems to assume the reader is already very familiar with CUDA specifically. That may not be great if the idea is to support non-CPU memory more generally.

Reply:
Same conversation going on here? https://github.com/apache/arrow/pull/34972/files#r1161080258

Reply:
resolving this in favor of the other conversation on the same topic: https://github.com/apache/arrow/pull/34972/files#r1161080258

Please unresolve and comment if this should be continued separately here for some reason.